app/controllers/lti_controller.rb

```diff
@@ -25,31 +26,125 @@ def get_canvas_config
               {
                 text: I18n.t('lti.launch'),
                 placement: 'course_navigation',
-                target_link_uri: root_url,
+                target_link_uri: lti_redirect_login_url,
                 canvas_icon_class: 'icon-lti',
                 default: 'disabled',
                 visibility: 'admins',
-                windowTarget: '_blank'
+                windowTarget: '_blank',
+                custom_fields: {
+                  user_id: '$Canvas.user.id',
+                  course_id: '$Canvas.course.id',
+                  course_name: '$Canvas.course.name'
+                }
               }
             ]
           }
         }
       ],
-      public_jwk_url: lti_public_jwk_url
+      public_jwk_url: lti_public_jwk_url,
+      custom_fields: {
+        user_id: '$Canvas.user.id',
+        course_id: '$Canvas.course.id',
+        course_name: '$Canvas.course.name'
+      }
     }
+
     render json: config.to_json
   end
 
   def launch
-    # TODO: implement this function
-    render json: { state: 'success' }
+    if params[:client_id].blank? || params[:login_hint].blank? ||
+      params[:target_link_uri].blank? || params[:lti_message_hint].blank?
+      head :unprocessable_entity
+      return
+    end
+    nonce = rand(10 ** 30).to_s.rjust(30, '0')
+    session[:client_id] = params[:client_id]
+    session[:nonce] = nonce
```

Comment:
You can remove this

Reply:
Ignored purposefully because canvas does in fact send the nonce in the JWT, so we should check it